Existing user account not used when user has multiple Moodle accounts with same email address

peterkeijsers opened the pull request:
This fixes a bug that ocured when a users tried to login using SAML2. The conditions are:

- User has multiple Moodle accounts with the same email address (or other mapped idP field attributes that are being used to find and login the user in Moodle).

The condition was count($records == 1) so a user was being found only when there was exactly 1 user account found. This would make the whole reset() call on the next line redundant because there would always be exactly 0 or 1 records...

So I think this needs to be `if (count($records) > 0) {` so every scenario where the sql query returns multiple users from the Moodle database, the first one will always be used.

hdagheda committed:
WR #333356: Manually apply below patch due to dir change:

extlib-simplesaml: backport mods to extlib. 30632972e780ff5d8ccafb2c379864f9138d5390
Added support for IdP initiation flow #105 6ebf501a378eec10f5b268650e7399b906aeb63c
Fixed bug with Idp initiatied entityid ca343939b8c12951abcd533704ded45068a2ac99

Edited 04/10/23 for peer review of pull #774

hdagheda committed:
WR #333356: Manually apply below patch due to dir change:

extlib-simplesaml: backport mods to extlib. 30632972e780ff5d8ccafb2c379864f9138d5390
Added support for IdP initiation flow #105 6ebf501a378eec10f5b268650e7399b906aeb63c
Fixed bug with Idp initiatied entityid ca343939b8c12951abcd533704ded45068a2ac99

Edited 04/10/23 for peer review of pull #774

hdagheda committed:
WR #333356: Manually apply below patch due to dir change:

extlib-simplesaml: backport mods to extlib. 30632972e780ff5d8ccafb2c379864f9138d5390
Added support for IdP initiation flow #105 6ebf501a378eec10f5b268650e7399b906aeb63c
Fixed bug with Idp initiatied entityid ca343939b8c12951abcd533704ded45068a2ac99

Edited 04/10/23 for peer review of pull #774

skodak committed:
set SESSION->saml2idp during unsolicited logins

This makes the IdP known in saml_login_complete()
and tenant callback.

skodak committed:
set SESSION->saml2idp during unsolicited logins

This makes the IdP known in saml_login_complete()
and tenant callback.

sarahjcotton committed:
Issue #768: Remove unnecessary param being sent to sendSAML2AuthnRequest()

sarahjcotton committed:
Issue #768: Remove unnecessary param being sent to sendSAML2AuthnRequest()

sarahjcotton committed:
Issue #768: Remove unnecessary param being sent to sendSAML2AuthnRequest()

sarahjcotton committed:
Issue #768: Force de-prioritisation of BINDING_HTTP_ARTIFACT

sarahjcotton committed:
Issue #768: Force de-prioritisation of BINDING_HTTP_ARTIFACT

sarahjcotton committed:
Issue #768: Force de-prioritisation of BINDING_HTTP_ARTIFACT

sarahjcotton committed:
Issue #768: Fix unit test

peterkeijsers committed:
Fix bug where user account could not be found when user has multiple accounts with same email address

This fixes a bug that occured when a users tried to login using SAML2. The conditions are:

- User has multiple Moodle accounts with the same email address (or other mapped idP field attributes that are being used to find and login the user in Moodle).

The condition was count($records == 1) so a user was being found only when there was exactly 1 user account found. This would make the hole reset() call on the next line redundant because there would always be exactly 0 or 1 records...

So I think this needs to be count($records) > 0 so every scenario where the sql query returns multiple users from the Moodle database, the forst one will always be used.

danmarsden commented:
I'm leaning towards objecting towards this change... If the system has 2 user accounts for the same user - arbitrarily "choosing" one to use is IMO the wrong approach.

Can you provide a bit more information here on the reason your site has 2 active accounts with the same email? - is it an accidental duplication (which IMO should be fixed manually rather than hiding it in auth_saml2) - or is there some other case for multiple accounts with the same email?

peterkeijsers commented:
Hi Dan

Thank you for your quick response. I Agree that arbitrarily choosing the first account from the list of records is not ideal. This however was already the logic in your plugin (the line 88) but was never very relevant due to the condition always resulting in only 1 record.

**Some context on our case:**
However in Moodle it is possible through the setting "allowaccountssameemail" to allow for multiple accounts with the same email address. In my case however. The users could (in rare cases) have an LTI account, a manual account and a SAML2 account. These are edge cases however but they are real. And also some users have email addresses with uppercase and lowercase. Depending on the setting "auth_saml2 | tolower" this could also result in possible duplicate accounts.
The best solution would be to prevent double accounts. But in our case (and as a better fallback) I think it is better to arbitrarily choose the first available record than to not allow the user to login. Because the user then gets the message that "an account with this email address already excists.".
So I would prefer to log the user in with one of the matched email addresses and parallel to that try to minimize and cleanup any duplicate records (manually) by the site admin.

I hope this provides some better insights on my thoughts leading to this PR.

Thanks for looking into this.

danmarsden commented:
The way I read that existing code is - get all the user records that match, If we get more than 1 record - something went wrong and we can't match the user so don't let them login. - the reset call just simplifies the array returned into a single record so it can be used in the API call below it. There are probably other ways that the same code could be written but the end result is the same - if we get more than 1 user in the response, we haven't got a unique indentifier so don't log the user in.

My opinion is that existing code provides correct behavior and my vote is to decline this change. If we can't match a unique user with a unique identifier, the correct process should be to not let the user login.

But I'm happy for others to pitch in and disagree and try to convince me otherwise! :-)

brendanheywood commented:
> The users could (in rare cases) have an LTI account, a manual account and a SAML2 account. These are edge cases however but they are real. And also some users have email addresses with uppercase and lowercase. Depending on the setting "auth_saml2 | tolower" this could also result in possible duplicate accounts.

Before I read this issue I came up with this same scenario and I agree this is valid. What I think is the better middle ground is if the get_user function looks for users in 2 passes:
1) first it looks for an exact match of the user which also matches on the auth = saml2 - if it finds a single account then it uses it and away we go
2) if it can't find an account then it searches more widely for accounts where auth != saml2 AND anyauth is on

If anyauth is off then the second step would not run. 

If the first step finds two saml2 users which match then that should still be an error.

So I think we need to move the 'anyauth' out from the saml_login_complete into the user_extractor::get_user function

danmarsden commented:
Took me a bit to understand what Brendan was suggesting as an improvement there... (thanks Brendan)

If the user has 2 accounts - eg one account with the auth column in their record set to "saml2" and one account with the auth method set to "manual" and they both have the same email - it should choose the saml one and let the user login.

However - if the user has multiple accounts with the auth method set to "email" or "oauth" or something else - it should just recognise that there are multiple accounts that it can't match, preventing the user from being able to login.

So we should modify the code like he suggests.. First look to see if we can find one user account with auth = "saml2"  that matches the email -  if we find one record we use it. (if we find multiple saml2 accounts with that email we prevent login).

Then - after the initial search for auth_saml2 accounts, if "anyauth" is enabled, we use similar code to what's there right now - get all users that match the email, if we find one record, log them in, if we find more than one, don't let them login.

Does that make sense and would it meet your needs? - do you think you could modify the pull request to do something similar?

peterkeijsers commented:
Hi @danmarsden and @brendanheywood 

Thank you for all the thorough answers and critique. I agree that my current pull-request is not sufficient enough and could use some improvements. I made it too quick looking at it now. 
I think your last suggestion @danmarsden is a good one. That would solve my usecase and improves upon the login logic of this plugin.

I will create a new improved pull request with your suggestions. However due to my limited time the coming weeks it will take some time to complete and update this PR.

aneno-m-e committed:
Update post.twig to point to publicly available post files (#793)

And remove deprecated post.php

Co-authored-by: Noemie Ariste <noemieariste@catalyst.net.nz>

aneno-m-e committed:
Update post.twig to point to publicly available post files (#793)

And remove deprecated post.php

Co-authored-by: Noemie Ariste <noemieariste@catalyst.net.nz>

aneno-m-e committed:
Update post.twig to point to publicly available post files (#793)

And remove deprecated post.php

Co-authored-by: Noemie Ariste <noemieariste@catalyst.net.nz>

jwalits committed:
[#801] fix issue with IdP initiated login after upgrade to v2.0.5 (#802)

* [#801] fix issue with IdP initiated login after upgrade to v2.0.5

* [#801] restore comment for searching purposes

jwalits committed:
[#801] fix issue with IdP initiated login after upgrade to v2.0.5 (#802)

* [#801] fix issue with IdP initiated login after upgrade to v2.0.5

* [#801] restore comment for searching purposes

jwalits committed:
[#801] fix issue with IdP initiated login after upgrade to v2.0.5 (#802)

* [#801] fix issue with IdP initiated login after upgrade to v2.0.5

* [#801] restore comment for searching purposes

catalystfd committed:
auth_saml2 Ignore invalid idp request parameters

If the client sends us an idp request parameter that doesn't exist, we
simply ignore it, instead of setting it in the session.

This can happen if user's store an old link, becuase saml2 sends you to
default idp if it doesn't match, but the session variable still gets set
to the invalid nonexisting idp

ziegenberg committed:
PHP 8.2 Creation of dynamic property warnings

Fixes: #836

Signed-off-by: Daniel Ziegenberg <daniel@ziegenberg.at>

nathan-catalystit committed:
Support Moodle 4.5

jay-oswald committed:
bump simplesamlphp to 2.3.3

jay-oswald committed:
fix file permissions

jay-oswald committed:
adding after gitignore removal

jay-oswald committed:
bump simplesamlphp to 2.3.3 and lint

jay-oswald committed:
change mimimum version to 404 due to simplesamlphp needing php 8.1

jay-oswald committed:
remove broken tests

jay-oswald committed:
remove placeholder test

jay-oswald committed:
update readme versions

tuanngocnguyen committed:
Merge pull request #845 from catalyst/support_moodle_4.5

Support for Moodle 404+

jay-oswald committed:
Bump version.php requirements for 404

jay-oswald committed:
fix version number for 404

jay-oswald committed:
fix version description

tuanngocnguyen committed:
Merge pull request #848 from catalyst/jay-oswald-patch-1

Bump version.php requirements for 404

keevan committed:
fix: set the tempdirdefault on the setting itself instead of inside a lang string

danmarsden committed:
Merge pull request #851 from catalyst/fix-tempdir-default

fix: set the tempdirdefault on the setting itself instead of inside a lang string

gbarat87 committed:
Update readme with new simplesamlphp version

gbarat87 committed:
bump simplesamlphp to 2.3.5

dmitriim committed:
issue #855: add namespace to generator_test class

dmitriim committed:
issue #855: fix namespace for all tests

jay-oswald committed:
Merge pull request #856 from catalyst/issue855

issue #855: add namespace to test classes

gbarat87 committed:
Version Bump

jay-oswald committed:
Merge pull request #853 from catalyst/MOODLE_404_STABLE-simplesamlphp-update

Update simplesamlphp to 2.3.5

x-iy committed:
fix #860 - ci badge MOODLE_404_STABLE

tuanngocnguyen committed:
Merge pull request #861 from catalyst/issue860-404

fix #860 - ci badge MOODLE_404_STABLE

ScottVerbeek committed:
Moodle 4.5 deprecated print_error, use moodle_exception

- Final deprecation of `\print_error()`. Please use the `\moodle_exception` class instead.
  For more information see [MDL-74484](https://tracker.moodle.org/browse/MDL-74484)

danmarsden committed:
Merge pull request #865 from catalyst/print_error-deprecation-in-moodle-405

Moodle 4.5 deprecated print_error, use moodle_exception

danmarsden committed:
remove legacy callbacks.

danmarsden committed:
New 4.5 and higher branch,
update readme with list of currently supported branches.

danmarsden committed:
Allow Moodle to function as an IDP. (#820)

* Allow Moodle to function as an IDP.

* Adds IDP settings

* en/disable moodle as an IDP.
* Whitelist services by their <saml:Issuer>

* Populate IDP saml response with user profile fields.

* fix line ending

---------

Co-authored-by: Mark Webster <mark.webster@catalyst-eu.net>

danmarsden committed:
add t19 branch to readme.

Commit:
update phpunit mock method call

Commit:
add tearDown method to runtests

Commit:
make test data provider methods static

jay-oswald committed:
initial add of composer dependency

jay-oswald committed:
swtich to vendor autoload

jay-oswald committed:
add first patch

jay-oswald committed:
patch result

jay-oswald committed:
add other patches

jay-oswald committed:
add patches result

jay-oswald committed:
remove old .extlib version

jay-oswald committed:
add upgrading readme with basic how to upgrade, and patch

paulholden committed:
Remove third party reference to removed `.extlib` directory.

Fixes #885 - see 2bb85429a.

danmarsden committed:
Merge pull request #886 from paulholden/third-party-extlib

Remove third party reference to removed `.extlib` directory.

jay-oswald committed:
clean up .extlib refrences

jay-oswald committed:
add entire vendor folder

danmarsden committed:
Merge pull request #888 from catalyst/composer-fixes

Clean up from composer change

jay-oswald committed:
Bump version to release Composer changes

danmarsden committed:
Merge pull request #889 from catalyst/composer-release

Bump version to release Composer changes

danmarsden committed:
update readme and version with 5.0 requirements.

danmarsden committed:
Update README.md with new 401_STABLE branch.

jay-oswald committed:
Update 401 support readme

danmarsden committed:
Merge pull request #899 from catalyst/500-readme

Update 401 support readme

paulholden committed:
Fix Behat execute helper method arguments.

Upstream changes in MDL-86231 amended the method contract. Also
remove some old backwards compat code, given this version of the
plugin has explicit support only from 5.0.

Fixes #917.

danmarsden committed:
Merge pull request #918 from paulholden/behat-execute

Fix Behat execute helper method arguments.

danmarsden committed:
Update behat_auth_saml2.php

matthewhilton committed:
[#922] Dont escape url during connectivity test

matthewhilton committed:
[#919] Fix method args

matthewhilton committed:
misc: automatic code standards fixups

matthewhilton committed:
Merge pull request #921 from catalyst/fixbehatphpdoc

Update behat_auth_saml2.php

Peterburnett committed:
Merge pull request #924 from catalyst/922-fix-encoding-issue-500

[#922] Dont escape url during connectivity test

dasistwas committed:
Fix #928: User profile fields are not correctly created in response xml

danmarsden committed:
Merge pull request #927 from aneno-m-e/MOODLE_500_STABLE

auth_saml2 Ignore invalid idp request parameters

danmarsden committed:
Merge pull request #930 from Wunderbyte-GmbH/MOODLE_500_STABLE

Fix #928: User profile fields are not correctly created in response xml

peterkeijsers committed:
Merge branch 'catalyst:MOODLE_500_STABLE' into patch-1
